Charon/filetypes/CuraPackage.py

```diff
@@ -0,0 +1,124 @@
+        self._ensureRelationExists(virtual_path=material_path_alias, relation_type="material", origin="/package.json")
+        self._writeToAlias(material_path_alias, package_filename, material_data)
+
+    ##  Gets a list of paths to quality files in the package.
+    def getQualities(self) -> List[str]:
+        return self.listPaths("/qualities")
+
+    ##  Add a new quality file.
+    #   \param: quality_data The data of the quality file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addQuality(self, quality_data: bytes, package_filename: str) -> None:
+        quality_path_alias = "/qualities"
+        self._ensureRelationExists(virtual_path=quality_path_alias, relation_type="quality", origin="/package.json")
+        self._writeToAlias(quality_path_alias, package_filename, quality_data)
+
+    ##  Gets a list of paths to machine definition files in the package.
+    def getMachines(self) -> List[str]:
+        return self.listPaths("/definitions")
+
+    ##  Add a new machine definition file.
+    #   \param machine_data The data of the machine definition file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addMachine(self, machine_data: bytes, package_filename: str) -> None:
+        machine_path_alias = "/definitions"
+        self._ensureRelationExists(virtual_path=machine_path_alias, relation_type="definition", origin="/package.json")
+        self._writeToAlias(machine_path_alias, package_filename, machine_data)
+
+    ##  Gets a list of paths to plugins in the package.
+    def getPlugins(self) -> List[str]:
+        return self.listPaths("/plugins")
+
+    ##  Add a new plugin.
+    #   \param plugin_data The ZIP file containing the plugin files converted to bytes.
+    #   \param plugin_id The ID of the plugin within the package.
+    #   \raises FileNotFoundError If a required file is not in the plugin ZIP file, this error will be raised.
+    def addPlugin(self, plugin_data: bytes, plugin_id: str) -> None:
+        plugin_path_alias = "/plugins"
+        self._ensureRelationExists(virtual_path=plugin_path_alias, relation_type="plugin", origin="/package.json")
+        ignore_string = re.compile("|".join(self.PLUGIN_IGNORED_FILES))
+        required_paths = ["{}/plugin.json".format(plugin_id), "{}/__init__.py".format(plugin_id)]
+        paths_to_add = []  # type: List[str]
+
+        # Open the bytes as ZipFile and walk through all the files.
+        with ZipFile(io.BytesIO(plugin_data), "r") as zip_file:
+
+            # Find which files to add.
+            for zip_item in zip_file.filelist:
+                if ignore_string.search(zip_item.filename):
+                    continue
+                paths_to_add.append(zip_item.filename)
+
+            # Validate required files.
+            for required_path in required_paths:
+                if required_path not in paths_to_add:
```

Comment:
The `in` operation is not efficient for lists, can we use a set for the `paths_to_add` variable?

Reply:
The list will be relatively small, so i doubt you will even notice the optimisation.

Reply:
Fixed